slides/k8s/owners-and-dependents.md

```diff
@@ -0,0 +1,157 @@
+# Owners and dependents
+
+- Some objects are created by other objects
+
+  (example: pods created by replica sets, themselves created by deployments)
+
+- When an *owner* object is deleted, its *dependents* are deleted
+
+  (this is the default behavior; it can be changed)
+
+- We can delete a dependent directly if we want
+
+  (but generally, the owner will recreate another right away)
+
+- An object can have multiple owners
+
+---
+
+## Finding out the owners of an object
+
+- The owners are recorded in the field `ownerReferences` in the `metadata` block
+
+.exercise[
+
+- Let's start a replicated `nginx` deployment:
+  ```bash
+  kubectl run yanginx --image=nginx --replicas=3
+  ```
+
+- Once it's up, check the corresponding pods:
+  ```bash
+  kuebectl get pods -l run=yanginx -o yaml | head -n 25
+  ```
+
+]
+
+These pods are owned by a ReplicaSet named yanginx-xxxxxxxxxx.
+
+---
+
+## Listing objects with their owners
+
+- This is a good opportunity to try the `custom-columns` output!
+
+.exercise[
+
+- Show all pods with their owners:
+  ```bash
+  kubectl get pod -o custom-columns=\
+  NAME:.metadata.name,\
+  OWNER-KIND:.metadata.ownerReferences[0].kind,\
+  OWNER-NAME:.metadata.ownerReferences[0].name
+  ```
+
+]
+
+Note: the `custom-columns` option should be one long option (without spaces),
+so the lines should not be indented (otherwise the indentation will insert spaces).
+
+---
+
+## Deletion policy
+
+- When deleting an object through the API, three policies are available:
+
+  - foreground (API call returns after all dependents are deleted)
+
+  - background (API call returns immediately; dependents are scheduled for deletion)
+
+  - orphan (the dependents are not deleted)
+
+- When deleting an object with `kubectl`, this is selected with `--cascade`:
+
+  - `--cascade=true` deletes all dependent objects (default)
+
+  - `--cascade=false` orphans dependent objects
+
+---
+
+## What happens when an object is deleted
+
+- It is removed from the list of owners of its dependents
+
+- If, for one of these dependents, the list of owners becomes empty ...
+
+  - if the policy is "orphan", the object stays
+
+  - otherwise, the object is deleted
+
+---
+
+## Orphaning pods
+
+- We are going to delete the Deployment and Replica Set that we created
+
+- ... without deleting the corresponding pods!
```

Comment:
Is it worth explaining the kind of failure mode(s) that would lead to this happening in the wild?

Reply:
As far as I can tell, this _should not_ happen (except in case of extreme bugs in the API machinery and/or data corruption in etcd); my goal were:

- re-emphasize that pods are created by repica set that are created by deployments
- have a completely artificial way to create a bunch of orphan pods :-) 

Do you think this needs an extra note?

Reply:
I think when you're showing people how to clean up a weird situation, it might be useful to give a slight amount of attention to "how to tell if you have this" and/or "what could cause this". Otherwise, confusion!

Reply:
Right! I added a slide explaining the various scenarios leading to orphan resources (and emphasizing that it would generally be at the initiative of the user).

Reply:
This is great!